### PR TITLE
Tame shutdown procedure.

### PR DESCRIPTION
**Pull request by olafurpg**
The "fix" in #357 was totally broken. Lesson learned, don't write async
code late at night. This commit refactors the shutdown logic to ensure
the following

- shutdown actually responds, previously it hung forever.
- we stop sending LSP logs + status bar updates during shutdown.
- exit runs after shutdown completes.
- shutdown cleans up resources and does nothing else. Previously,
  the cancellation process was triggering unnecessary expensive
  computation like resolving/fetching Bloop jars when not using the
  embedded server.

I have manually validated in both vim and Atom that this refactoring
makes the server shutdown correctly. Previously, it never exited in Atom
since shutdown hung.

### CODE REVIEW

File: metals/src/main/scala/scala/meta/internal/metals/SilentLanguageClient.scala [hunk 1]
```diff
@@ -1,31 +1,30 @@
...
-class SilentClient extends MetalsLanguageClient {
-  override def metalsStatus(params: MetalsStatusParams): Unit =
-    ()
+/**
+ * A language client that ignores all requests.
+ *
+ * Used during shutdown procedure to ensure no status bar updates
+ * or log messages are published during shutdown.
+ */
+object SilentLanguageClient extends MetalsLanguageClient {
```
Comment (gabro) [inline metals/src/main/scala/scala/meta/internal/metals/SilentLanguageClient.scala]: NoopLanguageClient?

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala [hunk 2]
```diff
@@ -355,6 +347,9 @@ class MetalsLanguageServer(
   @JsonNotification("exit")
   def exit(): Unit = {
     shutdown()
```
Comment (gabro) [inline metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala]: Is this call idepmpotent? Calling shutdown then exit will go through shutdown twice.

File: metals/src/main/scala/scala/meta/internal/metals/SilentLanguageClient.scala (hunk 1 above)
Comment (olafurpg) [reply metals/src/main/scala/scala/meta/internal/metals/SilentLanguageClient.scala]: Done.

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala (hunk 2 above)
Comment (olafurpg) [reply metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala]: Yes it is, added a comment to explain why.

**Comment by olafurpg** [comment]
Fun fact, a language server that hangs on `shutdown` can stop Atom from quitting. I had to force quit Atom before this PR.

**Pull request merged**